olm/src/jsMain/kotlin/io/github/matrixkt/olm/PkDecryption.kt

```diff
@@ -0,0 +1,92 @@
+package io.github.matrixkt.olm
+
+import org.khronos.webgl.Uint8Array
+import kotlin.random.Random
+
+public actual class PkDecryption {
+    private val ptr: JsOlm.PkDecryption
+    public actual val publicKey: String
+
+    private constructor(ptr: JsOlm.PkDecryption, publicKey: String) {
+        this.ptr = ptr
+        this.publicKey = publicKey
+    }
+
+    private constructor(ptr: JsOlm.PkDecryption) {
+        this.ptr = ptr
+        this.publicKey = ""
+    }
+
+    public actual constructor(random: Random) {
+        ptr = JsOlm.PkDecryption()
+        try {
+            this.publicKey = ptr.generate_key()
+        } catch (e: Exception) {
+            clear()
+            throw e
+        }
+    }
+
+    public actual fun clear() {
+        ptr.free()
+    }
+
+    public actual val privateKey: ByteArray
+        get() {
+            val key: Uint8Array = ptr.get_private_key()
+            val t: String = key.toString()
+            return t.encodeToByteArray()
+        }
+
+    public actual fun decrypt(message: PkMessage): String {
+        return ptr.decrypt(message.ephemeralKey, message.mac, message.cipherText)
+    }
+
+    public actual fun pickle(key: ByteArray): String {
+        return ptr.pickle(key.toString())
+    }
+
+    public actual companion object {
+        public actual val publicKeyLength: Long
+            get() {
+                // There is no equivalent in matrix-org/olm so we need to do it this way
+                val pkdec = JsOlm.PkDecryption()
```

Comment:
This is being leaked in the success case.

Reply:
Also, if it's done this way, it should probably be cached.

Reply:
i was unable to get the cache working, would you point me to some sort of doc? Thanks

Reply:
i do not see why it is being leaked in success case. there are two free in the catch and the try block.

Reply:
caching works now